Prototypal inheritance (#230)
javascript-tutorial/pt.javascript.info: 1-js/08-prototypes/01-prototype-inheritance/article.md

# Herança Prototipada

> Na programação, nós sempre queremos extender algum comportamento.

Reviewer: ```suggestion
Na programação, é comum querermos extender algum comportamento.
```

> Por exemplo, nós temos um objeto `user` com suas propriedades e métodos, e queremos criar duas variantes `admin` e `guest` ligeiramente diferentes. Nós gostaríamos de reutilizar o que já temos no `user`, sem copiar/reimplementar seus métodos, apenas construir um novo objeto baseado nele.

Reviewer: ```suggestion
Por exemplo, nós temos um objeto `usuario` com suas propriedades e métodos, e queremos criar duas variantes `admin` e `convidado` ligeiramente diferentes. Nós gostaríamos de reutilizar o que já temos no `usuario`, sem copiar/reimplementar seus métodos, apenas construir um novo objeto baseado nele.
```

Herança Prototipada (*Prototypal inheritance*) é uma funcionalidade da linguagem que nos ajuda com isso.

## [[Prototype]]

No Javascript, os objetos possuem uma propriedade escondida especial `[[Prototype]]` (como é chamada na especificação), cujo valor é `null` ou referencia outro objeto. Esse objeto é chamado de "protótipo":

![prototype](object-prototype-empty.svg)

Quando lemos uma propriedade de um `object`, e ela não está presente, o Javascript automaticamente obtém seu valor do protótipo. Na programação, isso é chamado de "herança prototipada". Em breve nós vamos ver vários exemplos dessa herança, bem como funcionalidades legais construidas em cima disso.

A propriedade `[[Prototype]]` é interna e escondida, mas existem várias formas de configurar um valor para ela. 

Uma delas é usar o nome especial `__proto__`, dessa forma:

> ```js run

Reviewer: Traduzir o código

*/!*
};

// walk é obtido da cadeia de protótipos
longEar.walk(); // Animal caminha
alert(longEar.jumps); // true (vindo de rabbit)
```

![](proto-animal-rabbit-chain.svg)

Agora, se nós quisermos ler algo de `longEar` e ele não possuir, o Javascript vai procurar no `rabbit`, e depois no `animal`.

Só existem duas limitações:

> 1. As referências não podem ser circulares. O Javascript vai disparar um erro caso a gente tente atribuir `__proto__` em um círculo de objetos.

Reviewer: ```suggestion
1. As referências não podem ser circulares. O Javascript vai disparar um erro caso a gente tente atribuir `__proto__` ao seu próprio objeto, ou ao objeto que o utiliza, de maneira recursiva.
```

Mas note que `__proto__` *não é o mesmo* que a propriedade interna `[[Prototype]]`. Ele é um getter/setter para o `[[Prototype]]`. Mais tarde veremos situações nas quais isso importa, por ora vamos apenar manter isso em mente enquanto construimos nossa compreensão da linguagem Javascript.

A propriedade `__proto__` está um pouco ultrapassada, ela existe por motivos históricos. O Javascript moderno sugere que nós usemos as funções `Object.getPrototypeOf/Object.setPrototypeOf` no lugar, que também fazem get/set do protótipo. Nós também vamos cobrir essas funções mais tarde. 

A especificação diz que o `__proto__` só pode ser suportado por browsers. Mas o fato é que todos os ambientes, incluindo o lado do servidor ("server-side") suportam o `__proto__`, então podemos usá-lo tranquilamente.

Como a notação `__proto__` é um pouco mais intuitiva, vamos usá-la nos exemplos.
```

## A Escrita não usa o protótipo

O protótipo é usado apenas para leitura de propriedades.

As operações de escrita/deleção trabalham diretamente com o objeto.

No exemplo abaixo, nós criamos um método `walk` próprio para o `rabbit`:

> ```js run

Reviewer: Traduzir o restante do código

*!*
rabbit.walk = function() {
  alert("Coelho! Pula-pula!");
};
*/!*

rabbit.walk(); // Coelho! Pula-pula!
```

De agora em diante, chamar `rabbit.walk()` encontra o método imediatamente no objeto e o executa, sem usar o protótipo:

![](proto-animal-rabbit-walk-2.svg)

Métodos de acesso são uma exceção, porque a atribuição é feita por uma função setter. Assim, guardar um valor numa dessas propriedades é o mesmo que chamar uma função.

Por essa razão, `admin.fullName` funciona corretamente no código abaixo:

> ```js run

Reviewer: Traduzir o código

```

Na linha `(*)`, a propriedade `admin.fullName` tem um getter no protótipo `user`, então ele é chamado. E na linha `(**)` a propriedade tem um setter no protótipo, então ele é chamado.

## O valor do "this"

Uma pergunta interessante pode surgir no exemplo acima: qual o valor do `this` dentro de `set fullName(value)`? Onde as propriedades `this.name` e `this.surname` são escritas: dentro de `user` ou de `admin`?

A resposta é simples: o `this` não é nem um pouco afetado pelos protótipos.

**Não importa onde o método é encontrado: dentro de um objeto ou do seu protótipo. Ao chamar um método, `this` é sempre o objeto antes do ponto.**

Então, a chamada do setter `admin.fullName=` usa o `admin` como `this`, não o `user`.

Na verdade isso é super importante, porque pode ser que a gente tenha um objeto grande com vários métodos, sendo que outros objetos herdam ele. Quando os objetos herdeiros rodam os métodos herdados, eles vão modificar apenas seus próprios estados, não o estado do objeto grande.

Por exemplo, aqui `animal` representa um "armazenador de métodos", e `rabbit` faz uso deles.

A chamada de `rabbit.sleep()` configura `this.isSleeping` no objeto `rabbit`:

> ```js run

Reviewer: Traduzir o resto do código

};

let rabbit = {
  name: "Coelho branco",
  __proto__: animal
};

// modifica rabbit.isSleeping
rabbit.sleep();

alert(rabbit.isSleeping); // true
alert(animal.isSleeping); // undefined (não existe essa propriedade no protótipo)
```

A imagem do resultado fica:

![](proto-animal-rabbit-walk-3.svg)

Se nós tivéssemos outros objetos, como `bird`, `snake`, etc., herdando de `animal`, eles iriam ganhar acesso aos métodos de `animal`. Mas em cada chamada dos métodos, o `this` corresponderia ao próprio objeto (antes do ponto), calculado na hora da chamada, não ao `animal`. Dessa forma, quando escrevemos algum dado no `this`, ele vai ser armazenado dentro de cada objeto.

Como resultado, os métodos são compartilhados, mas o estado dos objetos não são.

## Loop for..in

O loop `for..in` também itera sobre as propriedades herdadas.

Por exemplo:

> ```js run

Reviewer: Traduzir o código

  let isOwn = rabbit.hasOwnProperty(prop);

  if (isOwn) {
    alert(`Nosso: ${prop}`); // Nosso: jumps
  } else {
    alert(`Herdado: ${prop}`); // Herdado: eats
  }
}
```

Aqui nós temos a seguinte cadeia de heranças: `rabbit` herda de `animal`, que herda de `Object.prototype` (por definição, pois `animal` é um objeto literal `{...}`), que tem apenas `null` acima:

![](rabbit-animal-object.svg)

Note que existe uma coisa curiosa. De onde o método `rabbit.hasOwnProperty` está vindo? Nós não definimos ele. Mas olhando para a cadeia, nós podemos ver que esse método é passado pelo `Object.prototype.hasOwnProperty`. Em outras palavras, ele é herdado.

...Mas por que `hasOwnProperty` não aparece no loop `for..in`, da mesma forma que `eats` e `jumps`, já que o `for..in` lista propriedades herdadas?

A resposta é simples: ele não é enumerável. Assim como todas as outras propriedades de `Object.prototype`, ele tem a flag `enumerable:false`. E `for..in` só lista aquelas que são enumeráveis. Aí está o porquê do resto das propriedades de `Object.prototype` não estarem listadas.

> ```smart header="Quase todas as outras chaves/métodos *obtém-valor* ignoram propriedades herdadas."

Reviewer: ```suggestion
```smart header="Quase todos os métodos que obtém valores a partir de chaves ignoram propriedades herdadas."
```

Quase todas as outras chaves/métodos *obtém-valor*, como `Object.keys`, `Object.values`, entre outros, ignoram propriedades herdadas.

Eles operam apenas no próprio objeto. Propriedades vindas do protótipo *não* são levadas em conta.
```

## Resumo

- No Javascript, todos os objetos possuem uma propriedade `[[Prototype]]` escondida que ou é um objeto ou é `null`.

> - Nós podemos usar `obj.__proto__` para acessá-lo (um histórico getter/setter, mas já existem alternativas, que veremos em breve).

Reviewer: ```suggestion
- Nós podemos usar `obj.__proto__` para acessá-lo (um getter/setter mantido historicamente, mas já existem alternativas, que veremos em breve).
```

  let isOwn = rabbit.hasOwnProperty(prop);

  if (isOwn) {
    alert(`Nosso: ${prop}`); // Nosso: jumps
  } else {
    alert(`Herdado: ${prop}`); // Herdado: eats
  }
}
```

Aqui nós temos a seguinte cadeia de heranças: `rabbit` herda de `animal`, que herda de `Object.prototype` (por definição, pois `animal` é um objeto literal `{...}`), que tem apenas `null` acima:

![](rabbit-animal-object.svg)

Note que existe uma coisa curiosa. De onde o método `rabbit.hasOwnProperty` está vindo? Nós não definimos ele. Mas olhando para a cadeia, nós podemos ver que esse método é passado pelo `Object.prototype.hasOwnProperty`. Em outras palavras, ele é herdado.

...Mas por que `hasOwnProperty` não aparece no loop `for..in`, da mesma forma que `eats` e `jumps`, já que o `for..in` lista propriedades herdadas?

A resposta é simples: ele não é enumerável. Assim como todas as outras propriedades de `Object.prototype`, ele tem a flag `enumerable:false`. E `for..in` só lista aquelas que são enumeráveis. Aí está o porquê do resto das propriedades de `Object.prototype` não estarem listadas.

```smart header="Quase todas as outras chaves/métodos *value-gettting* ignoram propriedades herdadas."

> Quase todas as outras chaves/métodos *value-gettting*, como `Object.keys`, `Object.values`, entre outros, ignoram propriedades herdadas.

Reviewer 2: ```suggestion
Quase todas as outras chaves/métodos *obtém-valor*, como `Object.keys`, `Object.values`, entre outros, ignoram propriedades herdadas.
```

Eles operam apenas no próprio objeto. Propriedades vindas do protótipo *não* são levadas em conta.
```

## Resumo

- No Javascript, todos os objetos possuem uma propriedade `[[Prototype]]` escondida que ou é um objeto ou é `null`.

> - Nós podemos usar `obj.__proto__` para acessá-lo (um getter/setter que é histórico, para o qual já temos alternativas, que veremos em breve).

Reviewer 2: ```suggestion
- Nós podemos usar `obj.__proto__` para acessá-lo (um histórico getter/setter, mas já existem alternativas, que veremos em breve).
```

- O objeto referenciado por `[[Prototype]]` é um chamado "protótipo" (*prototype*).
- Se nós queremos ler uma propriedade de `obj` ou chamar um método, e eles não existem, o Javascript tenta encontrá-los no protótipo.

> - Operações de escrever/deletar agem diretamente no objeto, elas não usam o protótipo (assumindo que a propriedade seja um dado, não um setter).

Reviewer 2: ```suggestion
- Operações de escrever/deletar agem diretamente no objeto, elas não usam o protótipo (assumindo que seja uma propriedade para dados, não um setter).
```

- Se chamarmos `obj.method()`, e o `method` é obtido do prototype, o `this` ainda referencia o `obj`. Portanto métodos sempre trabalham com o objeto atual, mesmo que eles sejam herdados.

> - O loop `for..in` itera tanto sobre as propriedades do objeto quanto sobre as propriedades herdadas. Todas as outras chaves/métodos *value-getting* operam apenas sobre o próprio objeto.

Reviewer 2: ```suggestion
- O loop `for..in` itera tanto sobre as propriedades do objeto quanto sobre as propriedades herdadas. Todas os outros métodos  para obter chaves/valores operam apenas sobre o próprio objeto.
```
